File: src/data/shipsystems/scripts/ai/Diableavionics_grazerAI.java
package data.shipsystems.scripts.ai;

import com.fs.starfarer.api.combat.*;
import org.lazywizard.lazylib.MathUtils;
import org.lazywizard.lazylib.combat.AIUtils;
import org.lwjgl.util.vector.Vector2f;

public class Diableavionics_grazerAI implements ShipSystemAIScript {
    private CombatEngineAPI engine;
    private ShipAPI ship;
    private ShipSystemAPI system;
    private float getAggroRange(ShipAPI enemy) {
        if (enemy == null) return 0f;
        return switch (enemy.getHullSize()) {
            case FRIGATE    -> 800f;
            case DESTROYER  ->  1000f;
            case CRUISER    -> 1200f;
            case CAPITAL_SHIP-> 1400f;
            default         -> 600f;
        };
    }

    @Override
    public void init(
            ShipAPI shipAPI,
            ShipSystemAPI shipSystemAPI,
            ShipwideAIFlags shipwideAIFlags,
            CombatEngineAPI combatEngineAPI
    ) {
        this.engine = combatEngineAPI;
        this.ship = shipAPI;
        this.system = shipSystemAPI;
    }

    @Override
    public void advance(float amount, Vector2f missileDangerDir, Vector2f collisionDangerDir, ShipAPI target) {
        float timer = 0f;
        if (engine == null || ship == null || system == null || engine.isPaused()) {
            return;
        }
        ShipAPI nearestEnemy= AIUtils.getNearestEnemy(ship);
        if (nearestEnemy==null){
            return;
        }
        float aggroRange = getAggroRange(nearestEnemy);
        float distance = MathUtils.getDistance(ship, nearestEnemy);
        if (ship.getSystem().getAmmo() >= 3 && !ship.areAnyEnemiesInRange()) {
            ship.giveCommand(ShipCommand.USE_SYSTEM, null, 0);
            return;
        }

        if (ship.getFluxTracker().getFluxLevel() >= 0.8f && ship.getSystem().getAmmo() >= 1) {
            if (ship.getShield() != null) {
                ship.getShield().toggleOff();
            }

            while (timer <= 3) {
                ship.blockCommandForOneFrame(ShipCommand.TOGGLE_SHIELD_OR_PHASE_CLOAK);
                timer++;
            }
            ship.giveCommand(ShipCommand.USE_SYSTEM, null, 0);
            ship.giveCommand(ShipCommand.TOGGLE_SHIELD_OR_PHASE_CLOAK,null,0);
        }
        if (ship.getSystem().getAmmo()>=3 && distance>aggroRange){
            ship.giveCommand(ShipCommand.USE_SYSTEM,null,1);
        }
    }
}

File: src/data/shipsystems/scripts/ai/Diableavionics_grazerAI.kt
package data.shipsystems.scripts.ai

import com.fs.starfarer.api.combat.*
import org.lwjgl.util.vector.Vector2f

class Diableavionics_grazerAI : ShipSystemAIScript {
    private var engine: CombatEngineAPI? = null
    private var ship: ShipAPI? = null
    private var system: ShipSystemAPI? = null
    override fun init(
        shipAPI: ShipAPI?,
        shipSystemAPI: ShipSystemAPI?,
        shipwideAIFlags: ShipwideAIFlags?,
        combatEngineAPI: CombatEngineAPI?
    ) {
        this.engine = combatEngineAPI
        this.ship = shipAPI
        this.system = shipSystemAPI
    }

    override fun advance(v: Float, vector2f: Vector2f?, vector2f1: Vector2f?, shipAPI: ShipAPI?) {
        var timer = 0f
        if (engine == null || ship == null || system == null || engine!!.isPaused) {
            return
        }
        if (ship!!.system.ammo >= 3 && !ship!!.areAnyEnemiesInRange()) {
            ship!!.giveCommand(ShipCommand.USE_SYSTEM, null, 0)
            return
        }
        if (ship!!.fluxTracker.fluxLevel >= 0.8f && ship!!.system.ammo >= 1) {
            ship!!.shield.toggleOff()
            while (timer <= 3) {
                ship!!.blockCommandForOneFrame(ShipCommand.TOGGLE_SHIELD_OR_PHASE_CLOAK)
                timer++
            }
            ship!!.giveCommand(ShipCommand.USE_SYSTEM, null, 0)
        } else {
            ship!!.blockCommandForOneFrame(ShipCommand.USE_SYSTEM)
        }
    }
}
